src/tracked_struct.rs

```diff
@@ -181,6 +186,60 @@ pub struct IdentityHash {
     hash: u64,
 }
 
+#[derive(Default, Debug)]
+pub(crate) struct IdentityMap {
+    // we use a non-hasher hashmap here as our key contains its own hash (in a sense)
+    // so we use the raw entry api instead to avoid the overhead of hashing unnecessarily
+    map: hashbrown::HashMap<Identity, Id, ()>,
+}
+
+impl Clone for IdentityMap {
+    fn clone(&self) -> Self {
+        Self {
+            map: self.map.clone(),
+        }
+    }
+    fn clone_from(&mut self, source: &Self) {
+        self.map.clone_from(&source.map);
+    }
+}
+
+impl IdentityMap {
+    pub(crate) fn insert(&mut self, key: Identity, id: Id) -> Option<Id> {
+        use hashbrown::hash_map::RawEntryMut;
+
+        let eq_module_hash = |k: &Identity| {
+            k.ingredient_index == key.ingredient_index && k.disambiguator == key.disambiguator
+        };
+        let entry = self.map.raw_entry_mut().from_hash(key.hash, eq_module_hash);
```

Comment:
consider inlining `eq_module_hash`?

Reply:
Oh, that was supposed to be `eq_modulo_hash`. I prefer keeping it outside as it documents what we check (also the formatting like this is more readable imo)